cmake/feature_test.cmake

```diff
@@ -0,0 +1,45 @@
+include(CheckCXXSourceCompiles)
+
+function(BEMAN_FEATURE_TEST FEATURE_TEST_EXP RESULT_VAR)
+    check_cxx_source_compiles(
+        "
+    #if (${FEATURE_TEST_EXP})
+    int main() {}
+    #else
+    #error
+    #endif
+    "
+        ${RESULT_VAR}
+    )
+endfunction()
+
+function(BEMAN_MIN_CXX_VERSION CXX_VERSION)
+    set(VERSION_MAP
+        11
+        201103L
+        14
+        201402L
+        17
+        201703L
+        20
+        202002L
+        23
+        202302L
+    )
+
+    list(FIND VERSION_MAP "${CXX_VERSION}" INDEX)
+    if(INDEX EQUAL -1)
+        message(FATAL_ERROR "Unkown required C++ version: ${CXX_VERSION}")
+    else()
+        math(EXPR INDEX "${INDEX} + 1")
+        list(GET VERSION_MAP ${INDEX} VERSION_DATE)
+    endif()
+
+    beman_feature_test("__cplusplus >= ${VERSION_DATE}" CXX_VERSION_MET)
+    if(NOT CXX_VERSION_MET)
+        message(
+            FATAL_ERROR
```

Comment:
We can merge this as is and see if anyone complains... but I could see someone wanting to have an option to override our CMake helper and try building anyway. We could provide an CMake variable of some sort that would turn this message into a less severe category like `WARNING` or even `NOTICE`. See all the options in the docs for the CMake message command:

https://cmake.org/cmake/help/latest/command/message.html

Reply:
Egh, that's probably a good idea especially considering detecting this on MSVC is problematic.